qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+            start_time=self.start_time,
             carrier_freqs=carrier_freqs,
             phases=phases,
-            drift_array=Array(drift_array),
         )
 
-    def envelope_value(self, t: float) -> Array:
-        """Evaluate the envelope.
 
-        Args:
-            t: time
+class SignalList:
+    """A list of ``Signal``s, with functionality for simultaneous evaluation."""
 
-        Returns:
-            Array: the signal envelope at time t
-        """
-        return self.envelope(t)
+    def __init__(self, signal_list: List[Signal]):
+        self.components = signal_list
 
-    def value(self, t: float) -> Array:
-        """Evaluate the full value of the VectorSignal.
+    def complex_value(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of complex value of components."""
+        return np.moveaxis(
+            Array([Array(sig.complex_value(t)).data for sig in self.components]), 0, -1
+        )
 
-        Args:
-            t (float): time
+    def __call__(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of all components."""
+        return np.moveaxis(Array([Array(sig(t)).data for sig in self.components]), 0, -1)
 
-        Returns:
-            Array: the value of the signal (including carrier frequencies)
-                      at time t
-        """
-        carrier_val = np.exp(t * self._im_angular_freqs + 1.0j * self.phases)
-        return self.envelope_value(t) * carrier_val
+    def flatten(self) -> "SignalList":
+        """Return a ``SignalList`` with each component flattened."""
+        flattened_list = []
+        for sig in self.components:
+            if isinstance(sig, SignalSum):
+                flattened_list.append(sig.flatten())
+            else:
+                flattened_list.append(sig)
 
-    def conjugate(self):
-        """Return a new VectorSignal that is the complex conjugate of self.
+        return SignalList(flattened_list)
 
-        Returns:
-            VectorSignal: the complex conjugate of self
-        """
-        return VectorSignal(
-            lambda t: np.conjugate(self.envelope_value(t)),
-            -self.carrier_freqs,
-            -self.phases,
-            np.conjugate(self.drift_array),
+    @property
+    def drift(self) -> Array:
+        """Return the 'drift' Array, i.e. the constant part of the ``SignalList``."""
+
+        drift_array = []
+
+        for sig_entry in self.components:
+            val = 0.0
+
+            if not isinstance(sig_entry, SignalSum):
+                sig_entry = SignalSum(sig_entry)
+
+            for term in sig_entry:
+                if isinstance(term, Constant):
+                    val += term._value.data
+
+            drift_array.append(val)
+
+        return Array(drift_array)
+
+    def __getitem__(self, idx: Union[int, List, np.array, slice]) -> Signal:
+        """Enables numpy-style subscripting, as if this class were a 1d array."""
```

Comment:
This looks like a copy of `__getitem__` from `SignalSum`. It might make sense to introduce an abstract base class to remove some of this redundency.

Reply:
Yeah possibly - we could abstract out `__len__` as well. I'll think a bit, but it's probably worth it just for `__getitem__` alone since it's so long.